core/ordersync_subprotocols.go

```diff
@@ -2,99 +2,117 @@ package core
 	Page       int    `json:"page"`
 	SnapshotID string `json:"snapshotID"`
 }
 
-// Name returns the name of the FilteredPaginationSubProtocol
-func (p *FilteredPaginationSubProtocol) Name() string {
+// Name returns the name of the FilteredPaginationSubProtocolV0
+func (p *FilteredPaginationSubProtocolV0) Name() string {
 	return "/pagination-with-filter/version/0"
 }
 
 // HandleOrderSyncRequest returns the orders for one page, based on the page number
 // and snapshotID corresponding to the given request. This is
 // the implementation for the "provider" side of the subprotocol.
-func (p *FilteredPaginationSubProtocol) HandleOrderSyncRequest(ctx context.Context, req *ordersync.Request) (*ordersync.Response, error) {
-	var metadata *FilteredPaginationRequestMetadata
+func (p *FilteredPaginationSubProtocolV0) HandleOrderSyncRequest(ctx context.Context, req *ordersync.Request) (*ordersync.Response, error) {
+	var metadata *FilteredPaginationRequestMetadataV0
 	if req.Metadata == nil {
 		// Default metadata for the first request.
-		metadata = &FilteredPaginationRequestMetadata{
+		metadata = &FilteredPaginationRequestMetadataV0{
 			Page:       0,
 			SnapshotID: "",
 		}
 	} else {
 		var ok bool
-		metadata, ok = req.Metadata.(*FilteredPaginationRequestMetadata)
+		metadata, ok = req.Metadata.(*FilteredPaginationRequestMetadataV0)
 		if !ok {
-			return nil, fmt.Errorf("FilteredPaginationSubProtocol received request with wrong metadata type (got %T)", req.Metadata)
+			return nil, fmt.Errorf("FilteredPaginationSubProtocolV0 received request with wrong metadata type (got %T)", req.Metadata)
 		}
 	}
 
+	// Note(albrow): This version of Mesh does not support database snapshots. Instead, we use the SnapshotID
+	// field as minOrderHash.
+	var currentMinOrderHash common.Hash
+	if metadata.SnapshotID != "" {
+		if err := validateHexHash(metadata.SnapshotID); err != nil {
+			return nil, fmt.Errorf("FilteredPaginationSubProtocolV0 could not decode snapshotID (%q) as hex: %s", metadata.SnapshotID, err.Error())
+		}
+		currentMinOrderHash = common.HexToHash(metadata.SnapshotID)
+	}
+
 	// It's possible that none of the orders in the current page match the filter.
 	// We don't want to respond with zero orders, so keep iterating until we find
 	// at least some orders that match the filter.
 	filteredOrders := []*zeroex.SignedOrder{}
-	var snapshotID string
-	currentPage := metadata.Page
+	var nextMinOrderHash common.Hash
 	for {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		default:
 		}
 		// Get the orders for this page.
-		ordersResp, err := p.app.GetOrders(currentPage, p.perPage, metadata.SnapshotID)
+		ordersResp, err := p.app.GetOrders(p.perPage, currentMinOrderHash)
 		if err != nil {
 			return nil, err
 		}
-		snapshotID = ordersResp.SnapshotID
 		if len(ordersResp.OrdersInfos) == 0 {
 			// No more orders left.
 			break
 		}
+		nextMinOrderHash = ordersResp.OrdersInfos[len(ordersResp.OrdersInfos)-1].OrderHash
```

Comment:
What do you mean by this?

> nextMinOrderHash > currentMinOrderHash + perPage (where we convert these to numbers)

It doesn't make sense to add `perPage` to `currentMinOrderHash` since they are different types (`common.Hash` and `int`) and mean different things.

Reply:
As discussed offline, I was assuming that the hash would be cast into a number type (like `big.Int`). This suggestion is tracked by #822.